### PR TITLE
Add experimental dashboard to monitor simulations

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

This takes the `status` data and displays it in a web browser for all running jobs on the cluster, along with lots of plots to show the current state of BBH simulations. It's experimental, in that to use it you have to `pip install streamlit` by yourself and run `bin/python-spectre -m streamlit run tools/Status/Dashboard.py` to start the dashboard. There are some instructions in the file.

Here's a glance at what it looks like:

Bildschirmfoto 2024-06-29 um 14 04 37

I found this so useful when monitoring BBH simulations that I decided to open this experimental PR.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: tools/Status/Dashboard.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+            language="yaml",
+            line_numbers=True,
+        )
+
+    # Print outfile
+    with outfile_tab:
+        outfile = run_dir / "spectre.out"
+        if outfile.exists():
+            with open(outfile, "r") as open_outfile:
+                st.code("".join(["...\n"] + open_outfile.readlines()[-30:]))
```
Comment (knelli2) [inline tools/Status/Dashboard.py]: Any way to make this scrollable?

File: tools/Status/Dashboard.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    pages[username] = pages_user
+
+# Set up navigation
+page = st.navigation(pages)
+page.run()
+
+st.divider()
+
+# Refresh the page regularly
+if st.toggle("Auto-refresh", True):
```
Comment (knelli2) [inline tools/Status/Dashboard.py]: I found that if there is an error rending one of the plots of an executable status, then the auto-refresh is somehow disabled. Clicking on a different run brings it back. Any ideas on what to do for this?

File: tools/Status/ExecutableStatus/ExecutableStatus.py [hunk 3]
```diff
@@ -66,6 +66,25 @@ def format(self, field: str, value: Any) -> str:
         """
         raise NotImplementedError
 
+    def render_dashboard(self, job: dict, input_file: dict):
+        """Render a dashboard for the job (experimental).
+
+        Arguments:
+          job: A dictionary of job information, including the input file and
+            working directory. See the 'spectre.tools.Status.status' function
```
Comment (knelli2) [inline tools/Status/ExecutableStatus/ExecutableStatus.py]: `fetch_status` or `render_status` to be explicit?

File: tools/Status/ExecutableStatus/EvolveGhBinaryBlackHole.py [hunk 4]
```diff
@@ -94,3 +97,141 @@ def format(self, field, value):
...
+                        to_dataframe(
+                            open_h5file["ConstraintEnergy.dat"]
+                        ).set_index("Time")[["L2Norm(ConstraintEnergy)"]],
+                        to_dataframe(open_h5file["Norms.dat"]).set_index(
+                            "Time"
+                        )[
+                            [
+                                "L2Norm(PointwiseL2Norm(GaugeConstraint))",
+                                "L2Norm(PointwiseL2Norm(ThreeIndexConstraint))",
+                                "L2Norm(PointwiseL2Norm(FourIndexConstraint))",
```
Comment (knelli2) [inline tools/Status/ExecutableStatus/EvolveGhBinaryBlackHole.py]: Maybe use
```py
[col for col in df.columns if "Constraint" in col]
```
so that it will work if people stray from the pipeline a bit?

File: tools/Status/ExecutableStatus/EvolveGhBinaryBlackHole.py [hunk 5]
```diff
@@ -94,3 +97,141 @@ def format(self, field, value):
...
+                with st.expander("Size control A", expanded=False):
+                    st.pyplot(plot_size_control(reduction_files, "A"))
+                with st.expander("Size control B", expanded=False):
+                    st.pyplot(plot_size_control(reduction_files, "B"))
+
+        render_control_systems()
+
+        # Eccentricity
+        # (can be enabled once EccentricityControl is available)
+        # st.subheader("Eccentricity")
```
Comment (knelli2) [inline tools/Status/ExecutableStatus/EvolveGhBinaryBlackHole.py]: I think you've updated this section since you opened this. Wanna add those changes in?

File: tools/Status/ExecutableStatus/ExecutableStatus.py [hunk 6]
```diff
@@ -190,6 +202,51 @@ def format(self, field, value):
...
+        )
+        st.plotly_chart(fig)
+
+    def render_dashboard(self, job: dict, input_file: dict):
+        return self.render_time_steps(
+            input_file, list_reduction_files(job, input_file)
+        )
+
 
 class EllipticStatus(ExecutableStatus):
```
Comment (knelli2) [inline tools/Status/ExecutableStatus/ExecutableStatus.py:273]: I think you've also added a dashboard for the elliptic one too since you opened this. Wanna add that? I think the only thing is figuring out how to show the latest control step and not just the first one.

**Review by knelli2 — COMMENTED** [review]
These are all optional really. Just some things I noticed. We can't test any of this so once you feel it's ready we can just merge

File: tools/Status/Dashboard.py (hunk 1 above)
Comment (nilsvu) [reply tools/Status/Dashboard.py]: I tried loading the full file but that broke the server :D Added a comment

File: tools/Status/Dashboard.py (hunk 2 above)
Comment (nilsvu) [reply tools/Status/Dashboard.py]: Ah ok I moved the auto refresh further up in the script, that should fix it

File: tools/Status/ExecutableStatus/EvolveGhBinaryBlackHole.py (hunk 5 above)
Comment (nilsvu) [reply tools/Status/ExecutableStatus/EvolveGhBinaryBlackHole.py]: It's enabled now since #5895 was merged

File: tools/Status/ExecutableStatus/ExecutableStatus.py (hunk 6 above)
Comment (nilsvu) [reply tools/Status/ExecutableStatus/ExecutableStatus.py:273]: I added the elliptic dashboard. It only shows the convergence for the first control step atm, and also plots the convergence of the control parameter. I'll probably combine these somehow, but haven't decided how yet. Also AMR iterations are still missing from the status.

File: support/Pipelines/EccentricityControl/EccentricityControl.py [hunk 7]
```diff
@@ -120,10 +120,7 @@ def compute_separation(h5_file, subfile_name_aha, subfile_name_ahb):
         )
 
     # Compute separation
-    num_obs = len(ObjectA_centers[:, 0])
-
-    if len(ObjectB_centers[:, 0]) < num_obs:
-        num_obs = len(ObjectB_centers[:, 0])
+    num_obs = min(len(ObjectA_centers[:, 0]), len(ObjectB_centers[:, 0]))
```
Comment (knelli2) [inline support/Pipelines/EccentricityControl/EccentricityControl.py:123]: Maybe this commit and the one before should be in a separate PR since they are a bug fix? That way they don't get buried in an experimental PR

**Review by knelli2 — CHANGES_REQUESTED** [review]
Everything else looks ok

**Comment by knelli2** [comment]
@nilsvu  once the tests are fixed, we can merge this

**Comment by nilsvu** [comment]
Ok I squashed and rebased. Thanks for reviewing!

**Review by knelli2 — APPROVED** [review]
